### ISSUE TITLE
does not recognize exhaustive class patterns using `__match_args__` on generic union aliases

### ISSUE THREAD

**Issue by lennardwalter** [astral-sh/ty#3694]
### Summary

ty reports an implicit `None` return for an exhaustive `match` over a generic union alias when the final branch uses a nested class pattern such as `Err(str())`.
This seems to happen because ty treats the class pattern as refutable in isolation, but does not use the narrowed subject type plus `__match_args__` attribute mapping to prove that the nested pattern is exhaustive.

**Repro**
```py
class Ok[T]:
    __match_args__ = ("value",)
    value: T
class Err[E]:
    __match_args__ = ("value",)
    value: E
type Result[T, E] = Ok[T] | Err[E]
def f(result: Result[int, str]) -> int:
    match result:
        case Ok(value):
            return value
        case Err(str()):
            raise ValueError
```
**Actual**
ty reports:
```text
Function can implicitly return None
```
**Expected**
No diagnostic. After `case Ok(value)`, the remaining subject type is `Err[str]`, and `case Err(str())` should be exhaustive because `Err.__match_args__ = ("value",)` and `Err[str].value` is `str`.
**Possible Fix**
I have a simple local fix for this. It stores nested class-pattern arguments in the pattern predicate, resolves positional class-pattern arguments through exact literal `__match_args__`, and then checks whether each nested pattern is exhaustive for the matched attribute type of the already-narrowed subject.
It stays conservative: if `__match_args__` is dynamic or not known exactly, it does not claim exhaustiveness.

If this approach looks clean enough, I can open a PR. I needed this personally anyway, so I fixed it as minimally as I could. However as there is some match inference missing as far as I can tell, I would not expect this to be correct.

As

### Version

ty 0.0.44

**Comment by carljm** [comment]
Thanks for the report! I think this is expected given some known todo items in match pattern narrowing support (see #887). Will leave this open as a sub-issue so we can test our progress against this example.

**Comment by charliermarsh** [comment]
The exhaustiveness issue is fixed, and the code more broadly type-checks with `@final`:
```python
from typing import final

@final
class Ok[T]:
    __match_args__ = ("value",)
    value: T

@final
class Err[E]:
    __match_args__ = ("value",)
    value: E

type Result[T, E] = Ok[T] | Err[E]

def f(result: Result[int, str]) -> int:
    match result:
        case Ok(value):
            return value
        case Err(str()):
            raise ValueError
```
